### PR TITLE
feat(python): groupby_dynamic by quarter.

### LINKED ISSUE #6764: Include "quarter" as an option for `groupby_dynamic`.
### Problem description

I'd like to do `groupby_dynamic()` with business quarters, e.g. 
```python
df.groupby_dynamic("datetime_column_name", every="1q")  # every="1q" represents 'every business quarter', i.e. Jan-Mar, Apr-Jun, etc.
```

When I try this, I get the error `pyo3_runtime.PanicException: unit: 'q' not supported`. Either I can't figure out the correct code for "quarter", or the functionality isn't supported in polars.

For reference, [this use-case is supported by pandas](https://stackoverflow.com/questions/49949300/pandas-groupby-dates-within-a-quarter).

### PR DESCRIPTION
**Pull request by bfeif**
Closes #6764.

Note to Reviewer(s): this PR likely extends the same group-by-quarter functionality to other functions of the same ilk, e.g. `DataFrame.groupby_rolling()`, `Series.truncate()`, `Series.round()`, etc. Making the associated changes in the docstrings of all these functions is certainly possible, but would deviate from the scope of this PR. Ought these additional docstring changes be done in a separate Issue and PR, or in this PR? Please advise in the comments.

### CODE REVIEW

**Comment by ritchie46** [comment]
I don't think we should 3 months for `q`, but rather:

```
    First quarter, Q1: 1 January – 31 March (90 days or 91 days in leap years)
    Second quarter, Q2: 1 April – 30 June (91 days)
    Third quarter, Q3: 1 July – 30 September (92 days)
    Fourth quarter, Q4: 1 October – 31 December (92 days)
```

3 months is already available using `3mo`.

**Comment by bfeif** [comment]
@ritchie46 I agree with you, but it appears like that is the default behavior somehow. I've included a test to show what I'm talking about:

```python
def test_groupby_dynamic() -> None:
    import warnings
    dts = [
        datetime(2021, 12, 31, 0, 0, 0),
        datetime(2022, 1, 1, 0, 0, 1),
        datetime(2022, 3, 31, 0, 0, 1),
        datetime(2022, 4, 1, 0, 0, 1),
    ]
    df = (
        pl.DataFrame({
            "dt": dts
        })
    )
    out = (
        df
        .groupby_dynamic(
            "dt",
            every="1q"
        )
        .agg(
            pl.col("dt").agg_groups().alias("grouped_indices")
        )
    )
    warnings.warn(str(out))
```

The result of this code is:
```
  ┌─────────────────────┬─────────────────┐
  │ dt                  ┆ grouped_indices │
  │ ---                 ┆ ---             │
  │ datetime[μs]        ┆ list[u32]       │
  ╞═════════════════════╪═════════════════╡
  │ 2021-10-01 00:00:00 ┆ [0]             │
  │ 2022-01-01 00:00:00 ┆ [1, 2]          │
  │ 2022-04-01 00:00:00 ┆ [3]             │
  └─────────────────────┴─────────────────┘
```

In other words, the data is being grouped to quarters as per your explicit definition above. Any idea why this might be happening?

**Comment by ritchie46** [comment]
Because of where the data starts. It will not happen for all data. You must use the same approach as what we do with weeks and months.

**Comment by bfeif** [comment]
@ritchie46 I've tested it for a few different data start-dates, and it always groups to the correct quarter. Are you sure? To this point, I see that ["year" is defined as `"12mo"`](https://github.com/pola-rs/polars/blob/4afba2f905c0ff4cbbec095a0a1d3efae9fe38a4/polars/polars-time/src/windows/duration.rs#L135), and grouping by year has the desired behavior of actually grouping to the correct year. Why do you think that's the case for "year"?

**Comment by bfeif** [comment]
Bump -- Any thoughts on [this one](https://github.com/pola-rs/polars/pull/6842#issuecomment-1434284132) @ritchie46 ?

File: py-polars/tests/unit/test_datelike.py [hunk 1]
```diff
@@ -2343,3 +2343,29 @@ def test_infer_iso8601(iso8601_format: str) -> None:
...
+        df
+        .groupby_dynamic(
+            "dt",
+            every="1q"
+        )
+        .agg(
+            pl.col("dt").agg_groups().alias("grouped_indices")
+        )
+    )
+    warnings.warn(str(out))
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/test_datelike.py:2371]: can you assert the result against expected output?

File: py-polars/polars/internals/dataframe/frame.py [hunk 2]
```diff
@@ -3765,6 +3765,7 @@ def groupby_dynamic(
         - 1d    (1 day)
         - 1w    (1 week)
         - 1mo   (1 calendar month)
+        - 1q    (1 quarter)
```
Comment (MarcoGorelli) [inline py-polars/polars/internals/dataframe/frame.py:3768]: can you put this in all the relevant docs where duration strings are listed?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
I think this is probably fine, it works because by default there is `start_by='window'`, so by default it will split into 3-month periods starting in January

It's true that `'3mo'` already works, but it's not necessarily obvious to users, and it would be more familiar to those using `.dt.to_period('Q')` in pandas

If there's no objections, I'd suggest moving forwards with this one

**Comment by bfeif** [comment]
Thanks for the feedback @MarcoGorelli. I'm getting to work on your comments 👌

File: py-polars/polars/internals/dataframe/frame.py (hunk 2 above)
Comment (bfeif) [reply py-polars/polars/internals/dataframe/frame.py]: done.

**Comment by ritchie46** [comment]
Alright, the changes look good. A format run, and it should be good to go. Thank you @bfeif

File: py-polars/tests/unit/test_datelike.py (hunk 1 above)
Comment (bfeif) [reply py-polars/tests/unit/test_datelike.py]: done.

**Comment by bfeif** [comment]
@MarcoGorelli @ritchie46 It's ready for review.

**Pull request merged**